File: 643-maximum-average-subarray-i/maximum-average-subarray-i.py
class Solution:

  def findMaxAverage(self, nums: list[int], k: int) -> float:
    # Compute initial window sum
    window_sum = sum(nums[:k])
    max_sum = window_sum

    # Iterate with sliding window
    for i in range(k, len(nums)):
      window_sum += nums[i] - nums[i - k]
      if window_sum > max_sum:
        max_sum = window_sum

    return max_sum / k

File: 643-maximum-average-subarray-i/maximum-average-subarray-i.java
class Solution {
    public double findMaxAverage(int[] nums, int k) {
        // Compute the sum of the first window of size k
        double windowSum = 0;
        for (int i = 0; i < k; i++) {
            windowSum += nums[i];
        }

        double maxSum = windowSum;

        // Slide the window across the rest of the array
        for (int i = k; i < nums.length; i++) {
            windowSum += nums[i] - nums[i - k];
            maxSum = Math.max(maxSum, windowSum);
        }

        return maxSum / k;
    }
}